server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java

```diff
@@ -628,10 +632,7 @@ private void parseValueAddFields(ParseContext context, String value, String fiel
             }
 
             if (fieldType().hasDocValues()) {
-                if (context.doc().getField(fieldType().name()) == null || !context.doc().getFields(fieldType().name()).equals(field)) {
-                    if (fieldName.equals(fieldType().name())) {
```

Comment:
In the parent field, flat_object field, when fieldName.equals(fieldType().name()), we don't need to store the SortedSetDocValueField, because the parent field stores only the key of the field name in the JSON object, it's used in checking `exist query`. 
But in other queries, the mapping will direct the query to its subfields, either VALUE field or VALUE_AND_PATH field. 

The parent field type is not intended to used for sort, aggregate, and perform other operations on those values, will be directed to subfield instead, so  we can remove L632 to L634.

The expected length of indexable fields in a field {"field: { "foo": "bar"}}"

the parent field should contain one IndexableField, and the subfields should contains two IndexableField, in which one of them is SortedSetDocValueField: 
```
        IndexableField[] fields = doc.rootDoc().getFields("field");
        assertEquals(1, fields.length);
        assertEquals(new BytesRef("field.foo"), fields[0].binaryValue());

        IndexableFieldType fieldType = fields[0].fieldType();
        assertFalse(fieldType.tokenized());
        assertFalse(fieldType.stored());
        assertThat(fieldType.indexOptions(), equalTo(IndexOptions.DOCS));
        assertEquals(DocValuesType.NONE, fieldType.docValuesType());

        // Test internal substring fields as well
        IndexableField[] fieldValues = doc.rootDoc().getFields("field" + VALUE_SUFFIX);
        assertEquals(2, fieldValues.length);
        assertTrue(fieldValues[0] instanceof KeywordFieldMapper.KeywordField);
        assertEquals(new BytesRef("bar"), fieldValues[0].binaryValue());

        IndexableField[] fieldValueAndPaths = doc.rootDoc().getFields("field" + VALUE_AND_PATH_SUFFIX);
        assertEquals(2, fieldValues.length);
        assertTrue(fieldValueAndPaths[0] instanceof KeywordFieldMapper.KeywordField);
        assertEquals(new BytesRef("field.foo=bar"), fieldValueAndPaths[0].binaryValue());
```

Reply:
@reta @dblock @andrross @owaiskazi19 asking for opinions

Reply:
@mingshl The `FlatObjectFieldType` as of today (please correct me if I am wrong) is always created with `hasDocValues` set `true`, that would mean that `existsQuery` would use `DocValuesFieldExistsQuery` (which seems to be deprecated in flavor of `FieldExistsQuery`). If the parent field type is not intended is not intended to store doc values, we could instantiate it with `hasDocValues` set `false` instead.

Reply:
Agree that hasDocvalue is a bigger discussion. We should separate from this bug fix. I will raise a new issue related to the hasDocValue().